File: algorithms/ga.go
package algorithms

import (
	"math"
	"math/rand/v2"
	"sort"
	"ant_colony_golang_backend/models"
)

type GeneticAlgorithm struct {
	Points               []models.TurbineFaultPoint
	PopulationSize       int
	NGenerations         int
	NPoints              int
	MutationRate         float64
	ImplementLocalSearch bool
	Population           [][]int
	TurbineOrder         []string
	BestPath             []int
	BestPathLength       float64
	BestDowntimeDays     float64
	BestPathLenDowntime  float64
	Rnd                  *rand.Rand
}

func NewGeneticAlgorithm(points []models.TurbineFaultPoint, populationSize, nGenerations int, mutationRate float64, implementLocalSearch bool, rng *rand.Rand) *GeneticAlgorithm {
	if rng == nil {
		rng = rand.New(rand.NewPCG(42, 0))
	}
	if populationSize < 4 {
		populationSize = 4
	}
	ga := &GeneticAlgorithm{
		Points:               points,
		PopulationSize:       populationSize,
		NGenerations:         nGenerations,
		NPoints:              len(points),
		MutationRate:         mutationRate,
		ImplementLocalSearch: implementLocalSearch,
		BestPathLength:       math.Inf(1),
		BestDowntimeDays:     math.Inf(1),
		BestPathLenDowntime:  math.Inf(1),
		Rnd:                  rng,
	}
	ga.Population = ga.InitializePopulation()
	return ga
}

func (ga *GeneticAlgorithm) InitializePopulation() [][]int {
	if ga.NPoints == 0 {
		return [][]int{}
	}
	if ga.NPoints == 1 {
		pop := make([][]int, ga.PopulationSize)
		for i := 0; i < ga.PopulationSize; i++ {
			pop[i] = []int{0}
		}
		return pop
	}

	pop := make([][]int, ga.PopulationSize)
	for i := 0; i < ga.PopulationSize; i++ {
		ind := make([]int, ga.NPoints)
		for j := 0; j < ga.NPoints; j++ {
			ind[j] = j
		}
		ga.Rnd.Shuffle(ga.NPoints, func(a, b int) {
			ind[a], ind[b] = ind[b], ind[a]
		})
		pop[i] = ind
	}
	return pop
}

func (ga *GeneticAlgorithm) RouteCost(route []int) (float64, float64) {
	if len(route) <= 1 {
		return 0.0, 0.0
	}
	var fullPath []int
	if len(route) == ga.NPoints && ga.NPoints > 1 && route[0] != route[len(route)-1] {
		fullPath = make([]int, len(route)+1)
		copy(fullPath, route)
		fullPath[len(route)] = route[0]
	} else {
		fullPath = route
	}
	return ObjectiveFunction(fullPath, ga.Points)
}

func (ga *GeneticAlgorithm) CalculateTotalDistance(route []int) float64 {
	dist, _ := ga.RouteCost(route)
	return dist
}

type indEval struct {
	ind  []int
	cost float64
}

func (ga *GeneticAlgorithm) Selection() [][]int {
	evaluated := make([]indEval, len(ga.Population))
	for i, ind := range ga.Population {
		d, c := ga.RouteCost(ind)
		evaluated[i] = indEval{ind: ind, cost: d + c}
	}

	sort.Slice(evaluated, func(i, j int) bool {
		return evaluated[i].cost < evaluated[j].cost
	})

	numSelected := max(2, ga.PopulationSize/2)
	if numSelected > len(evaluated) {
		numSelected = len(evaluated)
	}
	selected := make([][]int, numSelected)
	for i := 0; i < numSelected; i++ {
		selected[i] = evaluated[i].ind
	}
	return selected
}

func (ga *GeneticAlgorithm) Crossover(parent1, parent2 []int) []int {
	if ga.NPoints <= 2 {
		res := make([]int, len(parent1))
		copy(res, parent1)
		return res
	}

	r1 := int(ga.Rnd.UintN(uint(ga.NPoints)))
	r2 := int(ga.Rnd.UintN(uint(ga.NPoints)))
	for r1 == r2 {
		r2 = int(ga.Rnd.UintN(uint(ga.NPoints)))
	}
	start, end := r1, r2
	if start > end {
		start, end = end, start
	}

	child := make([]int, ga.NPoints)
	inChild := make(map[int]bool)
	for i := range child {
		child[i] = -1
	}

	for i := start; i < end; i++ {
		child[i] = parent1[i]
		inChild[parent1[i]] = true
	}

	pointer := end % ga.NPoints
	for i := 0; i < ga.NPoints; i++ {
		gene := parent2[(end+i)%ga.NPoints]
		if !inChild[gene] {
			child[pointer] = gene
			inChild[gene] = true
			pointer = (pointer + 1) % ga.NPoints
		}
	}

	return child
}

func (ga *GeneticAlgorithm) Mutate(individual []int) []int {
	if ga.NPoints >= 2 && ga.Rnd.Float64() < ga.MutationRate {
		i := int(ga.Rnd.UintN(uint(ga.NPoints)))
		j := int(ga.Rnd.UintN(uint(ga.NPoints)))
		for i == j {
			j = int(ga.Rnd.UintN(uint(ga.NPoints)))
		}
		individual[i], individual[j] = individual[j], individual[i]
	}
	return individual
}

func (ga *GeneticAlgorithm) TwoOpt(route []int, maxIterations int) []int {
	if ga.NPoints < 4 {
		return route
	}

	best := make([]int, len(route))
	copy(best, route)
	d, c := ga.RouteCost(best)
	bestCost := d + c
	count := 0

	for count < maxIterations {
		improved := false
		for i := 0; i < ga.NPoints-1; i++ {
			for j := i + 1; j < ga.NPoints; j++ {
				if i == 0 && j == ga.NPoints-1 {
					continue
				}
				newRoute := make([]int, len(route))
				copy(newRoute[:i], best[:i])
				for k := 0; k <= j-i; k++ {
					newRoute[i+k] = best[j-k]
				}
				copy(newRoute[j+1:], best[j+1:])

				nd, nc := ga.RouteCost(newRoute)
				newCost := nd + nc
				if newCost < bestCost-1e-9 {
					best = newRoute
					bestCost = newCost
					improved = true
					break
				}
			}
			if improved {
				break
			}
		}
		if !improved {
			break
		}
		count++
	}
	return best
}

func (ga *GeneticAlgorithm) Evolve() {
	if ga.NPoints == 0 {
		ga.BestPath = []int{}
		ga.TurbineOrder = []string{}
		ga.BestPathLength = 0.0
		ga.BestDowntimeDays = 0.0
		ga.BestPathLenDowntime = 0.0
		return
	}

	if ga.NPoints == 1 {
		ga.BestPath = []int{0, 0}
		name := ga.Points[0].TurbineName
		if name == "" {
			name = "Point_0"
		}
		ga.TurbineOrder = []string{name, name}
		dist, dt := ObjectiveFunction(ga.BestPath, ga.Points)
		ga.BestPathLength = dist
		ga.BestDowntimeDays = dt
		ga.BestPathLenDowntime = dist + dt
		return
	}

	// Initial evaluation
	for _, individual := range ga.Population {
		closedInd := make([]int, len(individual)+1)
		copy(closedInd, individual)
		closedInd[len(individual)] = individual[0]

		dist, dtCost := ObjectiveFunction(closedInd, ga.Points)
		totalCost := dist + dtCost

		if totalCost < ga.BestPathLenDowntime {
			ga.BestPath = closedInd
			ga.BestPathLength = dist
			ga.BestDowntimeDays = dtCost
			ga.BestPathLenDowntime = totalCost

			order := make([]string, len(closedInd))
			for i, idx := range closedInd {
				order[i] = ga.Points[idx].TurbineName
			}
			ga.TurbineOrder = order
		}
	}

	for gen := 0; gen < ga.NGenerations; gen++ {
		selected := ga.Selection()
		newPop := make([][]int, 0, ga.PopulationSize)

		// Elitism: preserve top 2 individuals
		for i := 0; i < 2 && i < len(selected); i++ {
			preserved := make([]int, len(selected[i]))
			copy(preserved, selected[i])
			newPop = append(newPop, preserved)
		}

		for len(newPop) < ga.PopulationSize {
			p1Idx := int(ga.Rnd.UintN(uint(len(selected))))
			p2Idx := int(ga.Rnd.UintN(uint(len(selected))))
			for len(selected) > 1 && p1Idx == p2Idx {
				p2Idx = int(ga.Rnd.UintN(uint(len(selected))))
			}

			child := ga.Crossover(selected[p1Idx], selected[p2Idx])
			child = ga.Mutate(child)

			if ga.ImplementLocalSearch && ga.Rnd.Float64() < 0.2 {
				child = ga.TwoOpt(child, 10)
			}
			newPop = append(newPop, child)
		}
		ga.Population = newPop

		for _, individual := range ga.Population {
			closedInd := make([]int, len(individual)+1)
			copy(closedInd, individual)
			closedInd[len(individual)] = individual[0]

			dist, dtCost := ObjectiveFunction(closedInd, ga.Points)
			totalCost := dist + dtCost

			if totalCost < ga.BestPathLenDowntime {
				ga.BestPath = closedInd
				ga.BestPathLength = dist
				ga.BestDowntimeDays = dtCost
				ga.BestPathLenDowntime = totalCost

				order := make([]string, len(closedInd))
				for i, idx := range closedInd {
					order[i] = ga.Points[idx].TurbineName
				}
				ga.TurbineOrder = order
			}
		}
	}
}

File: rust/src/algorithms/ga.rs
use rand::seq::SliceRandom;
use rand::Rng;
use crate::models::TurbineFaultPoint;
use super::common::objective_function;

#[derive(Debug, Clone)]
pub struct GeneticAlgorithm {
    pub points: Vec<TurbineFaultPoint>,
    pub population_size: usize,
    pub n_generations: usize,
    pub n_points: usize,
    pub mutation_rate: f64,
    pub implement_local_search: bool,
    pub population: Vec<Vec<usize>>,
    pub turbine_order: Vec<String>,
    pub best_path: Vec<usize>,
    pub best_path_length: f64,
    pub best_downtime_days: f64,
    pub best_path_len_downtime: f64,
}

impl GeneticAlgorithm {
    pub fn new<R: Rng>(
        points: Vec<TurbineFaultPoint>,
        population_size: usize,
        n_generations: usize,
        mutation_rate: f64,
        implement_local_search: bool,
        rng: &mut R,
    ) -> Self {
        let pop_size = population_size.max(4);
        let n_points = points.len();

        let mut ga = Self {
            points,
            population_size: pop_size,
            n_generations,
            n_points,
            mutation_rate,
            implement_local_search,
            population: Vec::new(),
            turbine_order: Vec::new(),
            best_path: Vec::new(),
            best_path_length: f64::INFINITY,
            best_downtime_days: f64::INFINITY,
            best_path_len_downtime: f64::INFINITY,
        };

        ga.population = ga.initialize_population(rng);
        ga
    }

    pub fn initialize_population<R: Rng>(&self, rng: &mut R) -> Vec<Vec<usize>> {
        if self.n_points == 0 {
            return Vec::new();
        }
        if self.n_points == 1 {
            return vec![vec![0]; self.population_size];
        }

        let mut pop = Vec::with_capacity(self.population_size);
        for _ in 0..self.population_size {
            let mut ind: Vec<usize> = (0..self.n_points).collect();
            ind.shuffle(rng);
            pop.push(ind);
        }
        pop
    }

    pub fn route_cost(&self, route: &[usize]) -> (f64, f64) {
        if route.len() <= 1 {
            return (0.0, 0.0);
        }
        if route.len() == self.n_points && self.n_points > 1 && route[0] != route[route.len() - 1] {
            let mut full_path = Vec::with_capacity(route.len() + 1);
            full_path.extend_from_slice(route);
            full_path.push(route[0]);
            objective_function(&full_path, &self.points)
        } else {
            objective_function(route, &self.points)
        }
    }

    pub fn calculate_total_distance(&self, route: &[usize]) -> f64 {
        let (dist, _) = self.route_cost(route);
        dist
    }

    pub fn selection(&self) -> Vec<Vec<usize>> {
        let mut evaluated: Vec<(usize, f64)> = self
            .population
            .iter()
            .enumerate()
            .map(|(idx, ind)| {
                let (d, c) = self.route_cost(ind);
                (idx, d + c)
            })
            .collect();

        evaluated.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal));

        let num_selected = (self.population_size / 2).max(2).min(evaluated.len());
        let mut selected = Vec::with_capacity(num_selected);
        for &(idx, _) in &evaluated[..num_selected] {
            selected.push(self.population[idx].clone());
        }
        selected
    }

    pub fn crossover<R: Rng>(&self, parent1: &[usize], parent2: &[usize], rng: &mut R) -> Vec<usize> {
        if self.n_points <= 2 {
            return parent1.to_vec();
        }

        let r1 = rng.gen_range(0..self.n_points);
        let mut r2 = rng.gen_range(0..self.n_points);
        while r1 == r2 {
            r2 = rng.gen_range(0..self.n_points);
        }

        let (start, end) = if r1 < r2 { (r1, r2) } else { (r2, r1) };

        let mut child = vec![usize::MAX; self.n_points];
        let mut in_child = vec![false; self.n_points];

        for i in start..end {
            let gene = parent1[i];
            child[i] = gene;
            in_child[gene] = true;
        }

        let mut pointer = end % self.n_points;
        for i in 0..self.n_points {
            let gene = parent2[(end + i) % self.n_points];
            if !in_child[gene] {
                child[pointer] = gene;
                in_child[gene] = true;
                pointer = (pointer + 1) % self.n_points;
            }
        }

        child
    }

    pub fn mutate<R: Rng>(&self, individual: &mut [usize], rng: &mut R) {
        if self.n_points >= 2 && rng.gen::<f64>() < self.mutation_rate {
            let i = rng.gen_range(0..self.n_points);
            let mut j = rng.gen_range(0..self.n_points);
            while i == j {
                j = rng.gen_range(0..self.n_points);
            }
            individual.swap(i, j);
        }
    }

    pub fn two_opt(&self, route: &[usize], max_iterations: usize) -> Vec<usize> {
        if self.n_points < 4 {
            return route.to_vec();
        }

        let mut best = route.to_vec();
        let (d, c) = self.route_cost(&best);
        let mut best_cost = d + c;
        let mut count = 0;

        while count < max_iterations {
            let mut improved = false;
            for i in 0..self.n_points - 1 {
                for j in (i + 1)..self.n_points {
                    if i == 0 && j == self.n_points - 1 {
                        continue;
                    }

                    let mut new_route = Vec::with_capacity(self.n_points);
                    new_route.extend_from_slice(&best[..i]);
                    let mut reversed_segment = best[i..=j].to_vec();
                    reversed_segment.reverse();
                    new_route.extend(reversed_segment);
                    new_route.extend_from_slice(&best[j + 1..]);

                    let (nd, nc) = self.route_cost(&new_route);
                    let new_cost = nd + nc;
                    if new_cost < best_cost - 1e-9 {
                        best = new_route;
                        best_cost = new_cost;
                        improved = true;
                        break;
                    }
                }
                if improved {
                    break;
                }
            }
            if !improved {
                break;
            }
            count += 1;
        }

        best
    }

    pub fn evolve<R: Rng>(&mut self, rng: &mut R) {
        if self.n_points == 0 {
            self.best_path = Vec::new();
            self.turbine_order = Vec::new();
            self.best_path_length = 0.0;
            self.best_downtime_days = 0.0;
            self.best_path_len_downtime = 0.0;
            return;
        }

        if self.n_points == 1 {
            self.best_path = vec![0, 0];
            let name = if self.points[0].turbine_name.is_empty() {
                "Point_0".to_string()
            } else {
                self.points[0].turbine_name.clone()
            };
            self.turbine_order = vec![name.clone(), name];
            let (dist, dt) = objective_function(&self.best_path, &self.points);
            self.best_path_length = dist;
            self.best_downtime_days = dt;
            self.best_path_len_downtime = dist + dt;
            return;
        }

        // Initial evaluation
        for individual in &self.population {
            let mut closed_ind = Vec::with_capacity(individual.len() + 1);
            closed_ind.extend_from_slice(individual);
            closed_ind.push(individual[0]);

            let (dist, dt_cost) = objective_function(&closed_ind, &self.points);
            let total_cost = dist + dt_cost;

            if total_cost < self.best_path_len_downtime {
                self.best_path = closed_ind.clone();
                self.best_path_length = dist;
                self.best_downtime_days = dt_cost;
                self.best_path_len_downtime = total_cost;
                self.turbine_order = closed_ind
                    .iter()
                    .map(|&idx| self.points[idx].turbine_name.clone())
                    .collect();
            }
        }

        for _gen in 0..self.n_generations {
            let selected = self.selection();
            let mut new_pop = Vec::with_capacity(self.population_size);

            // Elitism: preserve top 2 individuals
            for item in selected.iter().take(2) {
                new_pop.push(item.clone());
            }

            while new_pop.len() < self.population_size {
                let p1_idx = rng.gen_range(0..selected.len());
                let mut p2_idx = rng.gen_range(0..selected.len());
                while selected.len() > 1 && p1_idx == p2_idx {
                    p2_idx = rng.gen_range(0..selected.len());
                }

                let mut child = self.crossover(&selected[p1_idx], &selected[p2_idx], rng);
                self.mutate(&mut child, rng);

                if self.implement_local_search && rng.gen::<f64>() < 0.2 {
                    child = self.two_opt(&child, 10);
                }
                new_pop.push(child);
            }

            self.population = new_pop;

            for individual in &self.population {
                let mut closed_ind = Vec::with_capacity(individual.len() + 1);
                closed_ind.extend_from_slice(individual);
                closed_ind.push(individual[0]);

                let (dist, dt_cost) = objective_function(&closed_ind, &self.points);
                let total_cost = dist + dt_cost;

                if total_cost < self.best_path_len_downtime {
                    self.best_path = closed_ind.clone();
                    self.best_path_length = dist;
                    self.best_downtime_days = dt_cost;
                    self.best_path_len_downtime = total_cost;
                    self.turbine_order = closed_ind
                        .iter()
                        .map(|&idx| self.points[idx].turbine_name.clone())
                        .collect();
                }
            }
        }
    }
}
